OpConnectSdk/Lib/Filter/FilterBuilder.cs

```diff
@@ -0,0 +1,185 @@
+using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Text;
+
+namespace OpConnectSdk.Lib.Filter
+{
+    public class FilterBuilder<T> {
+        private List<string> _tokens;
+
+        public FilterBuilder()
+        {
+            _tokens = new List<string>();
+        }
+
+        public FilterBuilder<T> Field<V>(Expression<Func<T, V>> fieldSelector)
+        {
+            Type type = typeof(T);
+
+            MemberExpression member = fieldSelector.Body as MemberExpression;
+            if (member == null)
+            {
+                throw new ArgumentException(string.Format("Expression '{0}' refers to a method, not a property.", fieldSelector.ToString()));
+            }
+
+            PropertyInfo propInfo = member.Member as PropertyInfo;
+            if (propInfo == null)
+            {
+                throw new ArgumentException(string.Format("Expression '{0}' refers to a field, not a property.", fieldSelector.ToString()));
+            }
+
+            if (type != propInfo.ReflectedType && !type.IsSubclassOf(propInfo.ReflectedType))
+            {
+                throw new ArgumentException(string.Format("Expression '{0}' refers to a property that is not from type {1}.", fieldSelector.ToString(), type));
+            }
+
+            var csharpFieldName = propInfo.Name;
+            if (csharpFieldName == null || csharpFieldName.Length == 0)
+            {
+                throw new ArgumentException(string.Format("Failed to retrieve property name from lambda: '{0}'", fieldSelector.ToString()));
+            }
+
+            // make the first character lowercase for the SCIM filter
+            var scimFieldName = csharpFieldName.Length > 1 ? char.ToLower(csharpFieldName[0]) + csharpFieldName.Substring(1) : csharpFieldName.ToLower();
+            _tokens.Add(scimFieldName);
+
+            return this;
+        }
+
+        public FilterBuilder<T> Group()
+        {
+            _tokens.Add("(");
+            return this;
+        }
+
+        public FilterBuilder<T> GroupEnd()
+        {
+            _tokens.Add(")");
+            return this;
+        }
+
+        public FilterBuilder<T> And()
+        {
+            _tokens.Add(ScimConstants.AND);
+            return this;
+        }
+
+        public FilterBuilder<T> Or()
+        {
+            _tokens.Add(ScimConstants.OR);
+            return this;
+        }
+
+        public FilterBuilder<T> Eq(string value)
+        {
+            _tokens.Add(ScimConstants.EQUALS);
+            _tokens.Add(value);
+            return this;
+        }
+
+        public FilterBuilder<T> Eq(long value)
+        {
+            _tokens.Add(ScimConstants.EQUALS);
+            _tokens.Add(value.ToString());
```

Comment:
These values still need to be wrapped in `" "`. So for instance, this would work: `_tokens.Add($"\"{value}\"")`. Or possibly they can all be done once in `ToString()`.